sdk/queue/priority_queue.go

```diff
@@ -0,0 +1,206 @@
+		return nil, ErrEmpty
+	}
+	item := heap.Pop(&pq.data).(*Item)
+	delete(pq.dataMap, item.Key)
+	return item, nil
+}
+
+// Push pushes an item on to the queue. This is a wrapper/convenience
+// method that calls heap.Push, so consumers do not need to invoke heap
+// functions directly. Items must have unique Keys, and Items in the queue
+// cannot be updated. To modify an Item, users must first remove it and re-push
+// it after modifications
+func (pq *PriorityQueue) Push(i *Item) error {
+	if i.Key == "" {
+		return errors.New("error adding item: Item Key is required")
+	}
+
+	pq.dataMutex.Lock()
+	defer pq.dataMutex.Unlock()
+
+	if _, ok := pq.dataMap[i.Key]; ok {
+		return ErrDuplicateItem
+	}
+
+	// copy the item value(s) so that modifications to the source item does not
+	// affect the item on the queue
+	clone, err := copystructure.Copy(i)
+	if err != nil {
+		return err
+	}
+
+	pq.dataMap[i.Key] = clone.(*Item)
+	heap.Push(&pq.data, clone)
+	return nil
+}
+
+// PopByKey searches the queue for an item with the given key and removes it
+// from the queue if found. Returns ErrItemNotFound(key) if not found. This
+// method must fix the queue after removal.
+func (pq *PriorityQueue) PopByKey(key string) (*Item, error) {
+	pq.dataMutex.Lock()
+	defer pq.dataMutex.Unlock()
+
+	item, ok := pq.dataMap[key]
+	if !ok {
+		return nil, newErrItemNotFound(key)
+	}
+
+	// remove the item the heap and delete it from the dataMap
+	itemRaw := heap.Remove(&pq.data, item.index)
+	delete(pq.dataMap, key)
+
+	if i, ok := itemRaw.(*Item); ok {
+		return i, nil
+	}
+
+	return nil, newErrItemNotFound(key)
+}
+
+// Len returns the number of items in the queue data structure. Do not use this
+// method directly on the queue, use PriorityQueue.Len() instead.
+func (pq pQueue) Len() int { return len(pq) }
+
+// Len returns the count of items in the Priority Queue
+func (q *PriorityQueue) Len() int { return q.data.Len() }
+
+// Less returns the less of two things, which in this case, we return the
+// highest value thing, because the priority ordering is "lowest value, highest
+// priority"
+func (pq pQueue) Less(i, j int) bool {
+	return pq[i].Priority < pq[j].Priority
+}
+
+// Swap swaps things in-place; part of sort.Interface
+func (pq pQueue) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].index = i
+	pq[j].index = j
+}
+
+// Push is used by heap.Interface to push items onto the heap. This method is
+// invoked by container/heap, and should not be used directly.
+// See: https://golang.org/pkg/container/heap/#Interface
+func (pq *pQueue) Push(x interface{}) {
+	n := len(*pq)
+	item := x.(*Item)
+	item.index = n
+	*pq = append(*pq, item)
+}
+
+// Pop is used by heap.Interface to pop items off of the heap. This method is
+// invoked by container/heap, and should not be used directly.
+// See: https://golang.org/pkg/container/heap/#Interface
+func (pq *pQueue) Pop() interface{} {
+	old := *pq
+	n := len(old)
+	item := old[n-1]
+	item.index = -1 //for saftey
```

Comment:
"safety"... but, what does this mean? What does setting index to -1 do?

Reply:
This code is copied verbatim from the example https://golang.org/pkg/container/heap/#example__priorityQueue  

I can only guess as to the purpose, perhaps to not give users a misleading sense of the internal indexing of the heap, or possibly pushing an item back on with an index from a prior time in heap could impact ordering if say a user forgets to call `Fix` or something, but I'm purely speculating.

Reply:
OK